oml/metrics/accumulation.py

```diff
@@ -75,20 +80,36 @@ def _put_in_storage(self, key: str, batch_value: Any) -> None:
         else:
             raise TypeError(f"Type '{type(batch_value)}' is not available for accumulating")
 
-    def update_data(self, data_dict: Dict[str, Any]) -> None:
+    def update_data(self, data_dict: Dict[str, Any], indices: Optional[List[int]] = None) -> None:
         """
         Args:
-            data_dict: We will accumulate data getting values via ``self.keys_to_accumulate``.
+            data_dict: We will accumulate data getting values via ``self.keys_to_accumulate``. All elements
+                       of the dictionary have to have the same size.
+            indices: Global indices of the elements in your batch of data. If provided, the accumulator
+                     will remove accumulated duplicates and return the elements in the sorted order after ``.sync()``.
+                     Indices may be useful in DDP (because data is gathered shuffled, additionally you may also get
+                     some duplicates due to padding). In the single device regime it's also useful if you accumulate
+                     data in shuffled order.
 
         """
-        bs_values = [len(data_dict[k]) for k in self.keys_to_accumulate]
+        keys = list(self.keys_to_accumulate)
+
+        if indices is not None:
+            assert isinstance(indices, List)
```

Comment:
Fortunately, this bug is caught by another assertion: we compare length of collected data and lengths of ids, but it happens only during the sync. So, I agree that we should raise an error earlier. 

I converted your example to a test case

Reply:
> collected data and lengths of ids
Not exactly. We use this method for assert:
```python
    def is_storage_full(self) -> bool:
        return self.num_samples == self.collected_samples
```
where `self.collected_samples` is incremented by `len` of each batch, not `indices`.

Anyway, with extra assertion it will work as expected.

Reply:
hm okay